File: downlink/python/anvil/tz.py
import datetime
import time


class tzoffset(datetime.tzinfo):

    #!defMethod(_,[seconds], [minutes], [hours])!2: "Create a timezone with a specific offset. Use an offset in seconds, minutes or hours" ["__init__"]
    def __init__(self, **kwargs):
        # Must initialise with seconds OR minutes OR hours, or with no arguments to get an offset of zero.
        if len(kwargs) > 1 or (
            len(kwargs) == 1 and "seconds" not in kwargs and "hours" not in kwargs and "minutes" not in kwargs
        ):
            raise TypeError(
                "tzoffset must be initialised with precisely one of 'seconds', 'minutes' or 'hours' keyword arguments"
            )

        self._offset = datetime.timedelta(**kwargs)

    def utcoffset(self, dt):
        return self._offset

    def dst(self, dt):
        return datetime.timedelta()

    def tzname(self, dt):
        return None

    def __repr__(self):
        mod = self.__module__
        name = type(self).__name__
        offset = self._offset.total_seconds() / 3600
        return "<{}.{} ({} hour offset)>".format(mod, name, offset)


#!defClass(anvil.tz, tzoffset)!:


# People should probably never use this, but we provide it for completeness.
class tzlocal(tzoffset):

    #!defMethod(_)!2: "Use the local timezone of the browser" ["__init__"]
    def __init__(self):
        if time.localtime().tm_isdst and time.daylight:
            s = -time.altzone
        else:
            s = -time.timezone
        tzoffset.__init__(self, seconds=s)


#!defClass(anvil.tz, tzlocal, anvil.tz.tzoffset)!:


class tzutc(tzoffset):

    #!defMethod(_)!2: "Create a timezone set to utc" ["__init__"]
    def __init__(self):
        tzoffset.__init__(self, minutes=0)

    def __repr__(self):
        return "<anvil.tz.tzutc>"


#!defClass(anvil.tz, tzutc, anvil.tz.tzoffset)!:

#!defModuleAttr(anvil.tz)!1: {name: 'UTC', type: 'any', description: 'An object representing the UTC timezone'}
UTC = tzutc()

File: client/js/modules/tz.ts
import {
    Args,
    Kws,
    buildPyClass,
    checkOneArg,
    lookupSpecial,
    pyCall,
    pyCallable,
    pyFunc,
    pyNewableType,
    pyNone,
    pyObject,
    pyStr,
    pyTypeError,
    toJs,
    toPy,
} from "@Sk";
import { PyModMap, datetimeMod, funcFastCall, kwsToObj } from "../runner/py-util";

/*#
id: tz_module
docs_url: /docs/server#advanced
title: Timezone Module
description: |
  ```python
  import anvil.tz
  ```

*/

const tz = () => {
    const pyMod: PyModMap = { __name__: new pyStr("anvil.tz") };

    // N.B. We assume that none of this suspends.
    const timedelta = datetimeMod.timedelta;
    const tzinfo = datetimeMod.tzinfo as pyNewableType;

    const s_offset = new pyStr("_offset");

    const init = (args: Args<[pyObject]>, kws: Kws = []) => {
        checkOneArg("tzoffset", args);
        const self = args[0];
        const kwObj = kwsToObj(kws);
        const kwLen = kws.length / 2;

        if (kwLen > 1 || !(kwLen === 0 || "hours" in kwObj || "minutes" in kwObj || "seconds" in kwObj)) {
            throw new pyTypeError(
                "tzoffset must be initialised with precisely one of 'seconds', 'minutes' or 'hours' keyword arguments"
            );
        }

        const offset = pyCall(timedelta, [], kws);
        self.tp$setattr(s_offset, offset);
        return pyNone;
    };

    const tzoffset = buildPyClass(
        pyMod,
        (_, $loc) => {
            $loc["__init__"] = funcFastCall(init);
            $loc["utcoffset"] = new pyFunc((self: pyObject) => self.tp$getattr(s_offset)!);
            $loc["dst"] = new pyFunc((_self: pyObject) => pyCall(timedelta));
            $loc["tzname"] = new pyFunc((_self: pyObject) => pyNone);
            $loc["__repr__"] = new pyFunc((self: pyObject) => {
                const totalSecondsFunc = self
                    .tp$getattr<pyCallable>(s_offset)
                    .tp$getattr<pyCallable>(toPy("total_seconds"));
                const modname = lookupSpecial(self, pyStr.$module);
                const name = self.tp$name;
                const totalSeconds = toJs(pyCall(totalSecondsFunc)) as number;
                return toPy(`<${modname}.${name} (${totalSeconds / 3600} hour offset)>`);
            });
        },
        "tzoffset",
        [tzinfo]
    );
    pyMod["tzoffset"] = tzoffset;

    const tzlocal = buildPyClass(
        pyMod,
        (_, $loc) => {
            $loc["__init__"] = new pyFunc((self: pyObject) =>
                init([self], ["minutes", toPy(-new Date().getTimezoneOffset())])
            );
            $loc["tzname"] = new pyFunc((_self: pyObject) => toPy("Browser Local"));
        },
        "tzlocal",
        [tzoffset as pyNewableType]
    );
    pyMod["tzlocal"] = tzlocal;

    const tzutc = buildPyClass(
        pyMod,
        (_, $loc) => {
            $loc["__init__"] = new pyFunc((self: pyObject) => init([self], ["minutes", toPy(0)]));
            $loc["tzname"] = new pyFunc((_self: pyObject) => toPy("UTC"));
            $loc["__repr__"] = new pyFunc((_self: pyObject) => toPy("<anvil.tz.tzutc>"));
        },
        "tzlocal",
        [tzoffset]
    );
    pyMod["tzutc"] = tzutc;

    pyMod["UTC"] = pyCall(tzutc as pyCallable);

    return pyMod;
};

export default tz;
